### ISSUE TITLE
Outbox pessimistic mode

### ISSUE THREAD

**Issue by DavidBoike** [Particular/NServiceBus.Persistence.Sql#37]
*Disclaimer: This idea is not fully baked.*

Currently Outbox does a SELECT for the Get stage, and does not insert until Store. This means that it's possible to have concurrently processing messages both enter the message handler, and while only one of the messages is able to commit its transaction successfully, the handler code still gets executed multiple times.

Of course this causes a problem if you're doing anything non-transactional. We could potentially decrease the likelihood that this happens by doing an INSERT in the Get stage. This should block the INSERT on the other thread from completing and returning to .NET code while the first handler completes. The second message would be prevented from entering the message handler.

Since it involves extra SQL statements, this would be slower, so it would need to be an opt-in option you could enable only on endpoints where you wanted to pay that price.

Of course each different engine would likely bring different wrinkles to an option like this, depending upon how they handle their transactions.

**Comment by SzymonPobiega** [comment]
@DavidBoike interesting thing. I am pretty sure I remember @ramonsmits proposing the same thing some time ago for NHibernate.

Well, in fact we could actually *replace* the `SELECT` with an `INSERT` because we want to optimize for non-duplicated messages.

**Comment by DavidBoike** [comment]
Yep, credit to @ramonsmits where credit is due. I looked for an issue for that in NHibernate but couldn't find it. He and I have had a conversation on the topic as well.

**Comment by weralabaj** [comment]
@weralabaj todo: look for similar issues in NHibernate for more context (high load on Outbox)

**Comment by SzymonPobiega** [comment]
I reviewed the code in the core and we can't do it (easily). The `OutboxTransaction` that ensures all changes in the outbox are done atomically starts *after* the initial lookup. If we change that lookup to be `INSERT` instead of `SELECT` and there is a failure down the line, we'll end up with a record but the message has not been processed.

**Comment by DavidBoike** [comment]
Well that's good to know. In light of that should we close this issue and raise a platform/core one instead?

**Comment by SzymonPobiega** [comment]
@DavidBoike I think this idea can be part of a broader "Make Outbox great again PlatDev issue"

**Comment by weralabaj** [comment]
@SzymonPobiega re: "make outbox great again" was there any progress in that area? is there an "umbrella" issue to refer to?

**Comment by SzymonPobiega** [comment]
@weralabaj I am not aware of it. I guess this could be a problem area if formulated correctly.

**Comment by weralabaj** [comment]
OK let's discuss in today's sync if we have enough time ;)

**Comment by weralabaj** [comment]
@SzymonPobiega Will verify if we can implement that in a minor under our budget, or need to escalate to PlatDev. Deadline tomorrow.

**Comment by SzymonPobiega** [comment]
OK, it seems this is doable. Here is the plan:

*Get*

```
SELECT * FROM OutboxData where MessageId = @MessageId
```

*BeginTransaction*

```
BEGIN TRAN
INSERT INTO OutboxData (MessageId) VALUES (@MessageId)
```

*Store*

```
UPDATE OutboxData SET Operations = @Operations WHERE MessageId = @MessageId
COMMIT TRAN
```

The insert in `BeginTransaction` would block one of the threads from actually processing the message. Once the first thread is done and calls `Store` the second thread will get an exception from SQL server that will cause the FLR. The outcome of such change will be:
 * We will not attempt to process a duplicate, saving resources
 * We will guarantee that a single message will not be handled concurrently on multiple threads

The cost of this change is an additional round-trip to the server. I am not sure how the driver works on the lower level but the calls to `insert` and `begin tran` could potentially be batched together on the wire?

This change would not resolve the problem with duplicate propagation because the loosing thread will attempt to re-process the message immediately (even sooner than in the current code) making it more likely to dispatch the message twice. For this to work we should use upgrade lock when calling `Get` and use two-phase `MarkAsDispatched` where in the first phase we update the record (before we actually dispatch the messages) and in the second phase we commit that transaction.

**Comment by DavidBoike** [comment]
So, in a minor?

**Comment by SzymonPobiega** [comment]
👍

**Comment by SimonCropp** [comment]
is this still happening for 3.1?

**Comment by SzymonPobiega** [comment]
I hope so

**Comment by SimonCropp** [comment]
given we r getting very close to code freeze and no one has worked on this i dont think it will make 3.1

**Comment by SzymonPobiega** [comment]
Fixed in #474
